### PR TITLE
Allow notifying about FS changes via the TAPI

### LINKED ISSUE #11594: Allow sending file changes to daemons via the tooling API
In order to allow IDEs to send changed files to all daemons which are watching for file system changes, so they pick up changes fast enough.

---
cc: @gradle/build-cache

### PR DESCRIPTION
**Pull request by wolfs**
This PR adds a new method to `ProjectConnection`, which sends a message
to all daemons and causes them to invalidate some part of the
retained file system state.

Fixes #11594

### CODE REVIEW

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import org.gradle.launcher.daemon.registry.DaemonInfo;
+import org.gradle.launcher.daemon.registry.DaemonRegistry;
+import org.gradle.launcher.daemon.server.api.DaemonStateControl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.UUID;
+
+public class DaemonFileSystemChangesNotificationClient {
```
Comment (lptr) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java]: Let's split this to interface and implementation for clarity.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    private final IdGenerator<UUID> idGenerator;
+    private final DaemonRegistry daemonRegistry;
+
+    public DaemonFileSystemChangesNotificationClient(DaemonConnector connector, IdGenerator<UUID> idGenerator, DaemonRegistry daemonRegistry) {
+        this.connector = connector;
+        this.idGenerator = idGenerator;
+        this.daemonRegistry = daemonRegistry;
+    }
+
+    public void notifyDaemonsAboutChangedFiles(List<String> locations) {
```
Comment (lptr) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java]: How about `notifyDaemonsAboutChangedPaths()`?

I think we should use the same term everywhere. Either "paths" or "file system locations".

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    private static void dispatch(Connection<Message> connection, Command stopCommand) {
+        Throwable failure = null;
+        try {
+            connection.dispatch(stopCommand);
+            Result result = (Result) connection.receive();
+            if (result instanceof Failure) {
+                failure = ((Failure) result).getValue();
+            }
+            connection.dispatch(new Finished());
+        } catch (Throwable e) {
```
Comment (lptr) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java]: Should we catch `Exception` here?

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/logging/DaemonMessages.java [hunk 4]
```diff
@@ -33,5 +33,6 @@
     public static final String ADVERTISING_DAEMON = "Advertising the daemon address to the clients: ";
     public static final String REMOVING_DAEMON_ADDRESS_ON_FAILURE = "Removing daemon from the registry due to communication failure. Daemon information: ";
     public static final String UNABLE_TO_STOP_DAEMON = "Unable to stop one of the daemons. The daemon may have crashed.";
+    public static final String UNABLE_TO_NOTIFY_DAEMON_ABOUT_FILE_CHANGES = "Unable to notify the daemon about file changes.";
```
Comment (lptr) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/logging/DaemonMessages.java]: If this is supposed to be sorted alphabetically then this one needs to go higher.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonServices.java [hunk 5]
```diff
@@ -142,6 +143,7 @@ protected GarbageCollectorMonitoringStrategy createGarbageCollectorMonitoringStr
         DaemonDiagnostics daemonDiagnostics = new DaemonDiagnostics(daemonLog, daemonContext.getPid());
         return ImmutableList.of(
             new HandleStop(get(ListenerManager.class)),
+            new HandleInvalidateLocations(this),
```
Comment (lptr) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonServices.java]: This should match the name of the command I think.

Comment (lptr) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonServices.java]: Do we have to pass `this`? That's always a smell...

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/protocol/InvalidateFileSystemLocations.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.launcher.daemon.protocol;
+
+import java.util.List;
+import java.util.UUID;
+
+public class InvalidateFileSystemLocations extends Command {
```
Comment (lptr) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/protocol/InvalidateFileSystemLocations.java]: We could also call this `InvalidateVirtualFileSystem` like we already have a cross-version test by this name.

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            apply plugin: 'java-library'
+        """
+    }
+
+    def cleanup() {
+        toolingApi.close()
+    }
+
+    def "can invalidate paths"() {
+        def filesToInvalidate = [file("src/main/java").absolutePath]
```
Comment (lptr) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy]: `pathsToInvalidate`? `locations`?

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+        when:
+        withConnection { connection ->
+            connection.model(GradleBuild).get()
```
Comment (lptr) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy]: Why is this part needed?

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java [hunk 9]
```diff
@@ -155,6 +156,13 @@
      */
     BuildActionExecuter.Builder action();
 
+    /**
+     * Notifies all daemons about the some changes made by an external tool, like an IDE.
```
Comment (lptr) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java]: Superfluous `the` (and potentially `some`). I think the point is that an external process made the changes, doesn't have to be a tool (whatever that means).

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultProjectConnection.java [hunk 10]
```diff
@@ -75,4 +80,28 @@ public TestLauncher newTestLauncher() {
     public BuildActionExecuter.Builder action() {
         return new DefaultBuildActionExecuter.Builder(connection, parameters);
     }
+
+    @Override
+    public void notifyDaemonsAboutChangedFiles(List<String> locations) {
+        ConsumerOperationParameters.Builder operationParamsBuilder = ConsumerOperationParameters.builder();
+        operationParamsBuilder.setCancellationToken(new DefaultCancellationTokenSource().token());
+        operationParamsBuilder.setParameters(parameters);
+        operationParamsBuilder.setEntryPoint("Invalidate locations API");
```
Comment (lptr) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultProjectConnection.java]: This should use the same terminology as all the rest.

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy (hunk 8 above)
Comment (wolfs) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy]: So I have a running daemon. If there isn't a daemon, nothing can be signalled. I'll add a comment.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java (hunk 1 above)
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java]: Why would that be clearer? `DaemonStopClient` is doing exactly the same.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java (hunk 3 above)
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java]: Good question. I am not sure how error handling should be done on the client/daemon in the tooling API. I copied this code from `StopDispatcher`.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/logging/DaemonMessages.java (hunk 4 above)
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/logging/DaemonMessages.java]: I added the new message at the end, since there seems to be no sorting in the file.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonServices.java (hunk 5 above)
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonServices.java]: I replaced `this` by `get(GradleUserHomeScopeServiceRegistry)`.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java (hunk 9 above)
Comment (wolfs) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java]: Fixed.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/protocol/InvalidateFileSystemLocations.java (hunk 6 above)
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/protocol/InvalidateFileSystemLocations.java]: Done.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java (hunk 2 above)
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/DaemonFileSystemChangesNotificationClient.java]: Using `paths` now consistently.

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy (hunk 7 above)
Comment (wolfs) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy]: `changedPaths`.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultProjectConnection.java (hunk 10 above)
Comment (wolfs) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultProjectConnection.java]: Done.

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/DefaultConnection.java [hunk 11]
```diff
@@ -42,7 +42,7 @@
 import org.gradle.tooling.internal.protocol.InternalCancellableConnection;
 import org.gradle.tooling.internal.protocol.InternalCancellationToken;
 import org.gradle.tooling.internal.protocol.InternalConnection;
-import org.gradle.tooling.internal.protocol.InternalFileSystemChangesAwareConnection;
+import org.gradle.tooling.internal.protocol.InternalNotifyDaemonsAboutChangedPathsConnection;
```
Comment (lptr) [inline subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/DefaultConnection.java]: This name is a bit weird. Not sure how to fix it. I like the ones that look like `InternalSomethingableConnection`.

Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/DefaultConnection.java]: How about `InternalInvalidatableVirtualFileSystemConnection`?

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+package org.gradle.integtests.tooling.r61
+
+import org.gradle.integtests.tooling.fixture.TargetGradleVersion
+import org.gradle.integtests.tooling.fixture.ToolingApiSpecification
+import org.gradle.integtests.tooling.fixture.ToolingApiVersion
+import org.gradle.tooling.model.gradle.GradleBuild
+
+@ToolingApiVersion(">=6.1")
+@TargetGradleVersion(">=6.1")
+class InvalidateVirtualFileSystemCrossVersionSpec extends ToolingApiSpecification {
```
Comment (big-guy) [inline subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy]: I think you should have a test that shows the behavior of trying to invalidate changes with an older version of Gradle.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/NotifyDaemonAboutChangedPathsClient.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        for (DaemonInfo daemonInfo : daemonRegistry.getAll()) {
+            DaemonStateControl.State state = daemonInfo.getState();
+            if (state != DaemonStateControl.State.Idle && state != DaemonStateControl.State.Busy) {
+                continue;
+            }
+            DaemonClientConnection connection = connector.maybeConnect(daemonInfo);
+            if (connection == null) {
+                continue;
+            }
+            dispatch(connection, new InvalidateVirtualFileSystem(changedPaths, idGenerator.generateId(), connection.getDaemon().getToken()));
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/NotifyDaemonAboutChangedPathsClient.java]: I am now only dispatching messages to idle daemons.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java [hunk 14]
```diff
@@ -155,6 +158,17 @@
...
+    /**
+     * Notifies all daemons about file changes made by an external process, like an IDE.
+     *
+     * The daemons will use this information to update the retained file system state.
+     *
+     * @param locations Absolute paths which have been changed by the external process.
+     * @since 6.1
+     */
+    @Incubating
+    void notifyDaemonsAboutChangedPaths(List<String> locations);
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java]: They can. What would be the advantage?

File: subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy (hunk 12 above)
Comment (wolfs) [reply subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r61/InvalidateVirtualFileSystemCrossVersionSpec.groovy]: I added the tests.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java (hunk 14 above)
Comment (adammurdoch) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java]: I think we should clarify a few things about the API, in the javadoc or perhaps by tweaking the signature:

- What kinds of exceptions are thrown?
- Is the operation synchronous or asynchronous? If asynchronous, how do I find out what happened?
- What do the paths mean? Are they absolute? relative? if so, to what?
- What should I pass when I rename something? delete something? create something?
- Are there any delivery guarantees if I start a build or some other operation after calling this method? that is, if I call this method and then immediately start a build, what happens?
- What happens if the version of Gradle I'm using does not support this feature?

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/NoToolingApiConnection.java [hunk 15]
```diff
@@ -64,4 +66,9 @@ public void runTests(TestExecutionRequest testExecutionRequest, ConsumerOperatio
         throw Exceptions.unsupportedFeature(operationParameters.getEntryPointName(), distribution, "2.6");
     }
 
+    @Override
+    public void notifyDaemonsAboutChangedPaths(List<String> changedPaths, ConsumerOperationParameters operationParameters) {
+        throw Exceptions.unsupportedFeature(operationParameters.getEntryPointName(), distribution, "6.1");
```
Comment (adammurdoch) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/NoToolingApiConnection.java]: should this fail? if it does, then we have a different contract for target versions of Gradle that support the tooling API but not the notification feature (ignore) and target versions of Gradle that don't support the tooling API at all (fail). On the other hand, there's almost no chance that this method will ever run anywhere except perhaps our test suite.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalNotifyDaemonsAboutChangedPathsConnection.java [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.tooling.internal.protocol;
+
+import java.util.List;
+
+public interface InternalNotifyDaemonsAboutChangedPathsConnection extends InternalProtocolInterface {
```
Comment (adammurdoch) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalNotifyDaemonsAboutChangedPathsConnection.java]: should add java docs similar to the others in this package.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalNotifyDaemonsAboutChangedPathsConnection.java [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+    void notifyDaemonsAboutChangedPaths(List<String> changedPaths, BuildParameters parameters);
```
Comment (adammurdoch) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalNotifyDaemonsAboutChangedPathsConnection.java]: I think it would be good to specify what exceptions this might fail with, as this is an important part of a cross version contract.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/NoToolingApiConnection.java (hunk 15 above)
Comment (wolfs) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/NoToolingApiConnection.java]: I think it should fail like the other methods. I can make the method no-op, though it seems like the user should be informed early that the version of the distribution is too low.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java [hunk 18]
```diff
@@ -155,6 +159,32 @@
...
+     * then the operation is a no-op.
+     *
+     * @param changedPaths Absolute paths which have been changed by the external process.
+     * @throws IllegalArgumentException When the paths are not absolute.
+     * @throws UnsupportedVersionException When the target Gradle version is &lt;= 2.5.
+     * @throws GradleConnectionException On some other failure using the connection.
+     * @since 6.1
+     */
+    @Incubating
+    void notifyDaemonsAboutChangedPaths(List<Path> changedPaths);
```
Comment (gavra0) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java]: Is the expectation that clients will invoke this only once, or this can be invoked many times (as soon as the file change is detected)?

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java [hunk 19]
```diff
@@ -155,6 +159,32 @@
...
+     *
+     * <p>The call is synchronous, i.e. the method ensures that the changed paths are taken into account
+     * by the daemon after it returned. This ensures that for every build started
+     * after this method has been called knows about the changed paths.
+     *
+     * <p>If the version of Gradle does not support virtual file system retention (i.e. &lt; 6.1),
+     * then the operation is a no-op.
+     *
+     * @param changedPaths Absolute paths which have been changed by the external process.
+     * @throws IllegalArgumentException When the paths are not absolute.
```
Comment (gavra0) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java]: Does it also throw is path is not canonical?

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java (hunk 18 above)
Comment (wolfs) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java]: This method should be invoked as soon as the file change is done, and shouldn't be invoked when the tool detect file changes by using a file watcher. So IDEA would invoke this method every time the user does something in the editor. I'll add a comment to the API.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java (hunk 19 above)
Comment (wolfs) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java]: Currently, it doesn't.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/NotifyDaemonAboutChangedPathsClient.java [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+    public NotifyDaemonAboutChangedPathsClient(DaemonConnector connector, IdGenerator<UUID> idGenerator, DaemonRegistry daemonRegistry) {
+        this.connector = connector;
+        this.idGenerator = idGenerator;
+        this.daemonRegistry = daemonRegistry;
+    }
+
+    public void notifyDaemonsAboutChangedPaths(List<String> changedPaths) {
+        for (DaemonInfo daemonInfo : daemonRegistry.getAll()) {
+            DaemonStateControl.State state = daemonInfo.getState();
+            if (state != DaemonStateControl.State.Idle) {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/NotifyDaemonAboutChangedPathsClient.java]: I created https://github.com/gradle/gradle/issues/11627 to track these improvements.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/NotifyDaemonAboutChangedPathsClient.java [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+            Result result = (Result) connection.receive();
+            if (result instanceof Failure) {
+                failure = ((Failure) result).getValue();
+            }
+            connection.dispatch(new Finished());
+        } catch (Throwable e) {
+            failure = e;
+        }
+        if (failure != null) {
+            LOGGER.warn(DaemonMessages.UNABLE_TO_NOTIFY_DAEMON_ABOUT_FILE_CHANGES, failure);
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/client/NotifyDaemonAboutChangedPathsClient.java]: Throwing now.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java (hunk 14 above)
Comment (wolfs) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java]: I clarified the Javadoc.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalNotifyDaemonsAboutChangedPathsConnection.java (hunk 16 above)
Comment (wolfs) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalNotifyDaemonsAboutChangedPathsConnection.java]: Added the Javadocs.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalNotifyDaemonsAboutChangedPathsConnection.java (hunk 17 above)
Comment (wolfs) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/InternalNotifyDaemonsAboutChangedPathsConnection.java]: Added.

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ProviderConnection.java [hunk 22]
```diff
@@ -182,6 +183,14 @@ public Object runTests(ProviderInternalTestExecutionRequest testExecutionRequest
         return run(action, cancellationToken, listenerConfig, listenerConfig.buildEventConsumer, providerParameters, params);
     }
 
+    public void notifyDaemonsAboutChangedPaths(List<String> changedPaths, ProviderOperationParameters providerParameters) {
+        LoggingServiceRegistry loggingServices = LoggingServiceRegistry.newNestedLogging();
+        Parameters params = initParams(providerParameters);
+        ServiceRegistry clientServices = daemonClientFactory.createStopDaemonServices(loggingServices.get(OutputEventListener.class), params.daemonParams);
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ProviderConnection.java]: Renamed.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java]: I am using `Path` now. Is that OK? I think we should use `Path` instead of `File` going forward.

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/GradleUserHomeScopeServiceRegistry.java [hunk 23]
```diff
@@ -35,6 +36,13 @@
      */
     ServiceRegistry getServicesFor(File gradleUserHomeDir);
 
+    /**
+     * Locates the shared services from the last used user home dir - if any.
+     *
+     * The method will never create services, so the caller should not release the registry.
+     */
+    Optional<ServiceRegistry> getCurrentServices();
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/service/scopes/GradleUserHomeScopeServiceRegistry.java]: I added this idea to https://github.com/gradle/gradle/issues/11627.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/protocol/InvalidateVirtualFileSystem.java [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.launcher.daemon.protocol;
+
+import java.util.List;
+import java.util.UUID;
+
+public class InvalidateVirtualFileSystem extends Command {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/protocol/InvalidateVirtualFileSystem.java]: Let's improve this in the next version, see https://github.com/gradle/gradle/issues/11627.

**Comment by wolfs** [comment]
@big-guy @adammurdoch: Please take another look.

The main change is that `Path` is now used as type in `ProjectConnection` and I added some validation that those `Path`s are actually absolute.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/logging/DaemonMessages.java [hunk 25]
```diff
@@ -34,4 +34,5 @@
     public static final String REMOVING_DAEMON_ADDRESS_ON_FAILURE = "Removing daemon from the registry due to communication failure. Daemon information: ";
     public static final String UNABLE_TO_STOP_DAEMON = "Unable to stop one of the daemons. The daemon may have crashed.";
     public static final String WAITING_ON_CANCELED = "Waiting for daemons with canceled builds to become available";
+    public static final String UNABLE_TO_NOTIFY_DAEMON_ABOUT_FILE_CHANGES = "Unable to notify the daemon about file changes.";
```
Comment (lptr) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/logging/DaemonMessages.java]: Nit: this should say "path" changes, no?

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/DefaultConnection.java (hunk 11 above)
Comment (lptr) [reply subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/DefaultConnection.java]: Sounds good!

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java (hunk 19 above)
Comment (lptr) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java]: I think it shouldn't throw if paths are not canonical. We might need to canonicalize, though, to handle symlinks properly.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultProjectConnection.java [hunk 26]
```diff
@@ -63,16 +70,47 @@ public TestLauncher newTestLauncher() {
...
     }
 
     @Override
     public BuildActionExecuter.Builder action() {
         return new DefaultBuildActionExecuter.Builder(connection, parameters);
     }
+
+    @Override
+    public void notifyDaemonsAboutChangedPaths(List<Path> changedPaths) {
+        List<String> absolutePaths = new ArrayList<>();
```
Comment (lptr) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultProjectConnection.java]: Let's allocate this to the size of `changedPaths`.

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/DefaultConnection.java (hunk 11 above)
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/DefaultConnection.java]: Renamed.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/logging/DaemonMessages.java (hunk 25 above)
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/logging/DaemonMessages.java]: I removed the constant since it is not used anymore - we now throw the exception.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultProjectConnection.java (hunk 26 above)
Comment (wolfs) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultProjectConnection.java]: Done.
